### DISCUSSION TITLE
Stop nether portal grief.

### DISCUSSION THREAD

**Discussion by zaphod77** [GriefPrevention/GriefPrevention#1876] [Ideas]
It's possible to trap players permanently inside a nether portal such that they can't use /trapped to escape.

Portal A is placed in your claim on overworld side, and surrounded by blocks.

Portal B in on nether side, and also surrounded by blocks

Portal C is outside your claim, but linked to Portal B

Player steps into portal c, and gets teleported to portal b. they are stuck there for 30 seconds, then get teleported to portal a, where they get stuck for 30 seconds, and get teleported to portal B where they get stuck for 30 seconds... and you can't enter chat during the nether portal animation.

If nether claims are allowed, it is 100% inescapable.

If they aren't, you still need a very powerful pick to break the frame on the nether side.

The server should immediately run a /trapped command on your behalf if you enter a nether portal, land in a claim, and are prevented from leaving the portal area.

**Comment by Jikoo** [comment]
GP does not initiate another nether teleport, it teleports to the original location. In your described situation, GP would never output the player to location A. They would return to location C after 30 seconds.

It is also difficult to determine if you are actually prevented from leaving the portal or lagging.

**Reply by zaphod77** [reply]
i actually couldn't leave.  the portal was blocked in. in my case portal b wasn't walled in, so i was able to escape

but you can still trap someone by claiming and quickly walling in portal b while they are stuck at portal c.

So the auto return teleport really needs to be an auto /trapped instead

**Reply by Jikoo** [reply]
That adds its own set of problems - GP's /trapped isn't exactly sophisticated and really was never designed to be used in the nether. I'm already imagining users on Purpur servers getting trapped and then killed by the ensuing void damage from a bit of careful setup after the "rescue." Frankly, there are so many edge cases for what GP could do with the player that you could practically write an entire plugin around it.

**Reply by Jikoo** [reply]
That said, feel free to submit a PR. My thoughts aren't Robo's, and as much as I have concerns about it, it does sound like a potential improvement.

**Reply by zaphod77** [reply]
well you'd run the command when on the non nether side, of course.

you go into nether, are trapped there for 30 seconds, but the griefer walls in the overworld side portal, so when the teleport is reverted, you are bounced back to a second trap.  At this point you are 100% screwed, and have to cheat to escape.

So i think if it only triggers in the overworld, it should be awfully hard to abuse.

If anyone else has a better way to prevent it i'm curious.

**Reply by Jikoo** [reply]
And when you enter a trap from the nether and the nether side is then trapped after you enter? What if neither portal is claimed? Your head should still be inside the unbreakable portal block, so you would not be able to break blocks to exit. It's not a bad idea, but there are definitely edge cases.

I think an ideal solution would be this:
1) Rescue player to portal entry point.
2) Search for sufficient passable blocks next to portal at entry point to exit, and, if found, do nothing.
   * Blocks should be within a short range - if you teleport into a 20-block-wide portal and there is a single 2-high gap at the other end, you can't reasonably expect to find it while the portal animation is going.
3) If player has build trust, destroy portal and display a warning. They can re-light it when safe.
   * Destroy all portal blocks within player's hitbox - player is not necessarily in the same block as the portal block, though in those cases they likely can destroy blocks to free themselves.
4) If player does not have build trust, rescue them out of the claim.
   * Rescue system should be slightly less garbage than it currently is so that players are not forced to very predictable and potentially unsafe locations - it's safe to assume that if griefers are manipulating the rescue system they're willing to manipulate further measures.
   * Alternately: Just destroy the portal anyway. Flint and steel charges are cheap, and people being inconvenienced for inconveniencing others seems fitting.

In general GP's rescue system is [not very good](https://github.com/TechFortress/GriefPrevention/blob/fe94717042415cbddfdf6cdaf52b01415a399794/src/main/java/me/ryanhamshire/GriefPrevention/GriefPrevention.java#L3303-L3330). It literally just moves 1 block from the lowest corner of the claim and loops until it finds an unclaimed block, then teleports you above the highest block at that coordinate. No guarantee of safety at all. You could appear in the middle of a massive lava lake for all GP cares, so a clever griefer could absolutely force that. I don't have any particularly effective ideas to improve it that aren't relatively resource-intensive. I think in an ideal world `/trapped` should rescue you to the last unclaimed location you were prior to entering a claim - if you managed safe entry to that point, you should be able to manage a safe exit. Unfortunately, that would likely involve the PlayerMoveEvent or some other similarly heavy check. That said, if other plugins can add entry and exit notifications reasonably efficiently, I don't see why GP can't track entry.

Also checked where this is in Mojira, and apparently it's an intended feature because allowing people to permanently trap others (or themselves) in nether portals is a far easier solution than bothering to actually fix the cause of dupe bugs with world transfer. You'd think they could special case the chat box at least, but from what I've seen as soon as something gets labelled WAI it's basically dead forever. https://bugs.mojang.com/browse/MC-2071

**Comment by zaphod77** [comment]
that's still better than permastuck.

Even if you forced it to drop you in the lava, it's still not permastuck.  

i have actually posted a suggestion or re-enabling guis after 10 seconds, under the assumption that if you haven't used the portal in ten seconds of wah wah wah then it's not going to happen until you leave the portal and reset the ten second counter. This is a 100% workable fix on minecraft side, and if i had access to the source, i'd try it.

Here are my suggested rules.

Use existing handling if stuck on nether side for 30 seconds.  send you back to where you were. this prevents the surprise back link trick.

If you are stuck in a portal for 30 seconds and it's in overworld and you are in a claim you can't build in, auto /trapped and grant you 1 minute of fire resistance to give you time to escape, either by /command or by swimming/mining/pillaring.

The rescue manipulation can be done without using nether portals ANYWAY.  Therefore that' a separate problem.  

But my simple  fix to /trapped is make it randomly select corner or middle of wall, randomly select one of the four corners/walls, and then randomly select the direction to search.
